Upstream gen spec script

ehildenb opened the pull request:
This script comes from the VSC repo originally.

It differs in that you can specify any level in the proof tree (not just the leaves), and it will generate all sub-proofs in the same spec file. For example, if your ini file is:

```
[pgm]
specs: transfer transferFrom
...

[root]
...

[transfer]
...

[transfer-success]
...

[transfer-success-1]
...

[transfer-failure]
...

[transferFrom]
....
```

You could say `./gen-spec.py DEFN_TMPL RULE_TMPL INI_FILE transfer-success-1 transfer-failer transferFrom` and it will generate one rule per spec for each of the listed proof names (so 3 specs total). But if you do `./gen-spec.py DEFN_TMPL RULE_TMPL INI_FILE transfer transferFrom`, it will only generate 2 specs, one with both `transfer-success-1` and `transfer-failure`, and another with `transferFrom`. If you don't supply any spec names, it defaults to the list in `pgm.specs`.

ehildenb committed:
k-distribution/bin/gen-specs.py: script for generating K specs from INI files and templates

ehildenb committed:
k-distribution/gen-specs: actually write out generated specs

ehildenb committed:
k-distribution/gen-spec: allow defaulting to list of specs in ini file

ehildenb committed:
k-distribution/gen-specs: default to outputting specs next to ini file

ehildenb committed:
k-distribution/gen-specs: more compact usage message

ehildenb committed:
k-distribution/gen-specs: correct number of arguments

ehildenb committed:
k-distribution/gen-specs: correctly read INI file

ehildenb commented:
Here is the KEVM PR which uses this new script: https://github.com/kframework/evm-semantics/pull/356

dwightguth commented:
I don't have time to look at the script in detail yet, but can we give it a better name, remove the .py, and add a wrapper in the distro-bin folder for it?

ehildenb committed:
k-distribution/gen-specs: allow specifying proof dependencies as trusted lemmas

ehildenb committed:
k-distribution/gen-specs: including trusted lemmas correctly

ehildenb committed:
k-distribution/gen-specs.py: allow case analysis and optimizations

ehildenb committed:
k-distribution/gen-specs: pass correct spec name to generate trusted

ehildenb committed:
k-distribution/gen-specs: shorter message

ehildenb committed:
k-distribution/{gen-specs.py => kprove_ini}: rename script

ehildenb committed:
k-distribution/{ktest.mak,kprove_ini}: integrate ini generation into makefile testing system

ehildenb committed:
k-distribution/ktest.mak: test ini file generation before running proofs

ehildenb committed:
regression-new/ini-spec-gen: add test of INI generated proofs

ehildenb committed:
regression-new/ini-spec-gen: one more spec

ehildenb committed:
regression-new/ini-spec-gen: update tests

ehildenb committed:
k-distribution/{ktest.mak,kprove_ini=>kprove-ini}: add distro-bin script

ehildenb commented:
@dwightguth I've renamed the script `gen-specs.py => kprove-ini`, and added a wrappen in `distro-bin` for the script.

I've also added a testing harness in `ktest.mak` which tests that the generated `*-spec.k` files are what we expect.

ehildenb committed:
k-distribution/kprove-ini: proper analysis of sub-cases of proofs

daejunpark committed:
regression-new/ini-spec-gen: add a new test seq
